docs/specs/app2app.md

```diff
@@ -0,0 +1,199 @@
+
+1. App B initiates an app2app authentication flow using an intent with uri `https://a.example.com/authorize?client_id=client_a&redirect_uri=https%3A%2F%2Fb.example.com%2Fredirect&code_challenge=CHALLENGE`.
+
+   - The parameters of the uri are:
+
+     - `client_id`: The client id of the initiating app.
+     - `redirect_uri`: The redirect uri used to recieve the authentication result.
+     - `code_challenge`: The PKCE code challenge with the same purpose in the oidc authorize endpoint.
+
+   - For security reason, App B should verify the signature of the app that is going to handle the intent by:
+     - Fetching the `.well-known/assetlinks.json` from the domain of the uri. In this example it will be `https://a.example.com/.well-known/assetlinks.json`.
+     - Obtain the package name and signature of the app which is going to handle the uri.
+     - Find the `"delegate_permission/common.handle_all_urls"` relation, and obtain the related `package_name` and `sha256_cert_fingerprints` from the section. Verify it with the package name and signature obtained in the previous step. If none of the items match, returns error.
+     - After validated the package name and signature, start an activity with an intent with the above uri.
+
+2. App A handles the intent.
+
+   - Returns an error if App A doesn't have an valid session.
+   - Find out the package name of the app which will handle the `redirect_uri`.
+   - Verify the signature of the above app by using the `.well-known/assetlinks.json` file, from the domain of the `redirect_uri`. In this example it is `https://b.example.com/.well-known/assetlinks.json`.
+   - Call the token endpoint with `grant_type=urn:authgear:params:oauth:grant-type:app2app`.
+     - `refresh_token` should be the refresh token of the existing session of App A.
+     - `client_id` should be the client id of this app.
+     - `app2app_client_id` should be obtained from the intent uri, which is the authorizing app.
+     - `challenge` should be generated with a private key stored in the android keystore, created on the app first authenticated with the server. Refer to the Server-side section for details.
+     - `app2app_redirect_uri` should be the redirect uri provided. The server will validate if it is a valid redirect uri.
+   - Create an intent with the redirect uri, with the following parameters appended to the uri:
+     - `code`: The new authorization code.
+     - In this example it should be `https://b.example.com/redirect?code=CODE`
+
+3. App B handles the redirect uri intent.
+
+   - Extract `code` from the uri and use it to perform the code exchange.
+
+#### Android SDK
+
+The following methods will be added in android sdk to support the app2app flow:
+
+- `startApp2AppAuthentication(authorizeUri: String, redirectUri: String)`
```

Comment:
I think the signature of this method should be

`(authorizationEndpoint: String, clientID: string, redirectURI: string)`

Example

`(authorizationEndpoint: "https://a.example.com/authorize", clientID: "client_a", redirectURI: "https://b.example.com/redirect")`

Then the developer do not need to compose the URI for us. In the previous version, we need to parse the partial composed URI and add query parameters.